uc-crux-llvm/src/UCCrux/LLVM/Classify.hs

```diff
@@ -0,0 +1,557 @@
+                 case getPtrOffsetAnn (Crucible.unRV ptr) of
+                   Just (Some (TypedSelector ftRepr (SomeInSelector (SelectArgument idx cursor)))) ->
+                     do
+                       let tag = ArgWriteBadAlignment
+                       liftIO $
+                         (appCtx ^. log) Hi $
+                           Text.unwords
+                             [ "Diagnosis:",
+                               diagnose tag,
+                               "#" <> Text.pack (show (Ctx.indexVal idx)),
+                               "at",
+                               Text.pack (show (ppCursor (argName idx) cursor)),
+                               "(" <> Text.pack (show (LLVMPtr.ppPtr (Crucible.unRV ptr))) <> ")"
+                             ]
+                       liftIO $ (appCtx ^. log) Hi $ prescribe tag
+                       case isPtrRepr ftRepr of
+                         Nothing -> panic "classify" ["Expected pointer type"]
+                         Just (IsPtrRepr Refl) ->
+                           return $
+                             ExMissingPreconditions
+                               (tag, oneArgConstraint idx cursor (Aligned alignment))
+                   _ -> unclass appCtx badBehavior
+             LLVMErrors.BBUndefinedBehavior
+               (UB.FreeUnallocated ptr) ->
+                 case getPtrOffsetAnn (Crucible.unRV ptr) of
+                   Just (Some (TypedSelector ftRepr (SomeInSelector (SelectArgument idx cursor)))) ->
+                     do
+                       let tag = ArgFreeUnallocated
+                       liftIO $
+                         (appCtx ^. log) Hi $
+                           Text.unwords
+                             [ "Diagnosis:",
+                               diagnose tag,
+                               "#" <> Text.pack (show (Ctx.indexVal idx)),
+                               "at",
+                               Text.pack (show (ppCursor (argName idx) cursor)),
+                               "(" <> Text.pack (show (LLVMPtr.ppPtr (Crucible.unRV ptr))) <> ")"
+                             ]
+                       liftIO $ (appCtx ^. log) Hi $ prescribe tag
+                       case isPtrRepr ftRepr of
+                         Nothing -> panic "classify" ["Expected pointer type"]
+                         Just (IsPtrRepr Refl) ->
+                           return $
+                             ExMissingPreconditions
+                               (tag, oneArgShapeConstraint idx cursor (Allocated 1))
+                   _ -> unclass appCtx badBehavior
+             LLVMErrors.BBUndefinedBehavior
+               (UB.FreeBadOffset (Crucible.RV ptr)) ->
+                 case getPtrOffsetAnn ptr of
+                   Just (Some (TypedSelector ftRepr (SomeInSelector (SelectArgument idx cursor)))) ->
+                     -- At the moment, we only handle the case where the pointer is
+                     -- unallocated.
```

Comment:
Are you asking about the `FreeBadOffset` classifier? It should be raised whenever `free` is called on a pointer that doesn't (necessarily) point to the base of an allocation. One reason that can happen (and this is the reason that's handled here) is that `free` was called on an unconstrained/unallocated/fully symbolic pointer that was produced by `Setup`. The fix is just to add a precondition that the pointer points to allocated memory.

In the future, this case should probably be expanded so that when this error happens on pointers that *aren't* from an argument, it's treated as a true positive.

Reply:
Does that answer your question? Is there a way the code/comment could be made more clear?

Reply:
I think that answers my question, thanks